Fix scatter batching rule for scatter_apply

jakevdp opened the pull request:
The issue is that the batching rule assumes that each scatter variant
always has the same update_jaxpr. This is not true of scatter_apply, which
lowers to scatter with a custom update_jaxpr. To address this, we change
the batching rule such that it re-uses the input jaxpr rather than always
re-generating it.

Fixes #16655

mattjj reviewed (approved):
Awesome work! Consider putting a line or two of the explanation you gave me in our 1:1 into the PR message :P

jakevdp committed:
Fix scatter batching rule for scatter_apply

The issue is that the batching rule assumes that each scatter variant
always has the same update_jaxpr. This is not true of scatter_apply, which
lowers to scatter with a custom update_jaxpr. To address this, we change
the batching rule such that it re-uses the input jaxpr rather than always
re-generating it.

jakevdp commented:
> Awesome work! Consider putting a line or two of the explanation you gave me in our 1:1 into the PR message :P

Done!
